### ISSUE TITLE
Using CBC Authenticated encryption modes with HMAC floods InMemoryCryptoProviderCache

### ISSUE THREAD

**Issue by Lilk** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1292]
The following code in AuthenticatedEncryptionProvider.cs loses tracks of the incoming SecurityKey's KeyId, and creates new keys. With Caching turned on their InternalKeyID (new Guid) is used as the cache key, which causes a memory leak. 
Could be solved by prepending or appending "AES" and "HMAC" respectively to the incoming KeyId, but that would opinionate the keyspace, preferred would likely be to hold separate ConcurrentDictionaries for these.

```C#
    private AuthenticatedKeys GetAlgorithmParameters(SecurityKey key, string algorithm)
        {
            int keyLength = -1;
            if (algorithm.Equals(SecurityAlgorithms.Aes256CbcHmacSha512, StringComparison.Ordinal))
                keyLength = 32;
            else if (algorithm.Equals(SecurityAlgorithms.Aes192CbcHmacSha384, StringComparison.Ordinal))
                keyLength = 24;
            else if (algorithm.Equals(SecurityAlgorithms.Aes128CbcHmacSha256, StringComparison.Ordinal))
                keyLength = 16;
            else
                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10668, GetType(), algorithm, key)));

            var keyBytes = GetKeyBytes(key);
            byte[] aesKey = new byte[keyLength];
            byte[] hmacKey = new byte[keyLength];
            Array.Copy(keyBytes, keyLength, aesKey, 0, keyLength);
            Array.Copy(keyBytes, hmacKey, keyLength);
            return new AuthenticatedKeys()
            {
                AesKey = new SymmetricSecurityKey(aesKey),
                HmacKey = new SymmetricSecurityKey(hmacKey)
            };
        }
```

**Comment by brentschmaltz** [comment]
@Lilk yes the keys are created without keyid so they can't be remembered or mapped to the original key.

Would you like to propose a fix as a PR?

**Comment by Lilk** [comment]
I guess I could see if I can come up with something :) Which branch would this be a PR to? dev? master?

**Comment by brentschmaltz** [comment]
@Lilk thanks!
we normally check into dev, then cherry pick as needed.

**Comment by Lilk** [comment]
@brentschmaltz better late than never, but I've a minimal "PR" which i think solves the issue. But I don't seem to have access to push a new branch and make a PR to dev from it. How do we solve this?

**Comment by brentschmaltz** [comment]
@Lilk could we use a https://tools.ietf.org/html/rfc7638?
To generate a kid based on the key material in a key.
The Guid model, was a quick fix.

**Comment by Lilk** [comment]
I can agree tp that the guid approach looks hacky ;).

Yes, as noted in the PR:

> We could probably assign a new deterministically generated Id based on the key content (such as a hash of the key content), as it would be more resilient against misuse of the library (Reuse of keyid for distinct keys)

Is RFC7638 implementation something that is WIP, or would you like a PR on that? ;)

**Comment by simonfox** [comment]
I assume this issue is related to the following issues:
https://github.com/Azure/azure-signalr/issues/647
https://github.com/Azure/azure-signalr/issues/654

I'm not using the signalr package, but am suffering from the issues described in there.

What is the current status of this issue? Seems to have gone quiet?

Also is the following the suggested workaround:
`CryptoProviderFactory.DefaultCacheSignatureProviders = false;`

**Comment by Lilk** [comment]
Yes, I believe it to be the same or highly related. We disabled the cache "temporarily". Had to focus on internal projects so I haven't been able to prioritise a PR for this issue. 
Personally I'm still uncertain to whether any work has gone into a RFC7638 implementation for cache key generation, or if such a PR would be appreciated by the maintainers.

**Comment by brentschmaltz** [comment]
@Lilk @simonfox we put in this fix: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1354 which allows for ignoring caching in such cases we should extend this logic to this case.

**Comment by brentschmaltz** [comment]
Fixed in 6.5.0 with: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1358
